psi4/src/psi4/libmints/zora.cc

```diff
@@ -0,0 +1,277 @@
+            auto pos_a = molecule_->xyz(a);
+
+            const double* coef_a  = &coeffs[c_aIndex[Z-1]];
+            const double* alpha_a = &alphas[c_aIndex[Z-1]];
+            int nc_a = c_aIndex[Z] - c_aIndex[Z-1];
+
+            //einsums("i,ip->p", 𝕔[i], erf(α[i]⊗ r[p]))/r[p]
+            for (int p = 0; p < npoints; p++) {
+                double dist = std::hypot(pos_a[0]-x[p], pos_a[1]-y[p], pos_a[2]-z[p]);
+                double outer = 0;
+                for (int i = 0; i < nc_a; i++) {
+                    outer += std::erf(dist * alpha_a[i]) * coef_a[i];
+                }
+                outer -= Z;
+                outer /= dist;
+                veff_block->add(p, outer);
+            }
+        }
+
+#pragma omp critical
+        {
+            veff_->insert({index, veff_block});
+        }
+    }
+}
+
+// Compute the scalar relativistic kinetic energy integral in the AO basis.
+// See header for equation.
+void ZORA::compute_TSR(std::vector<std::shared_ptr<BasisFunctions>> pworkers, SharedMatrix &T_SR) {
+	// Speed of light in atomic units squared
+	const double C2 = pc_c_au*pc_c_au;
+
+	int nthreads = 1;
+#ifdef _OPENMP
+    nthreads = Process::environment.get_n_threads();
+#endif
+
+	int max_points = grid_->max_points(); //Set in grid_int_options
+#pragma omp parallel num_threads(nthreads)
+	{
+		// Give each thread their own scratch space
+		std::vector<double> kernel(max_points);
+		auto tmp = std::make_shared<Matrix>(T_SR->ncol(), T_SR->nrow());
+#pragma omp for schedule(auto)
+	    for (const auto &block : grid_->blocks()) {
+			const auto &bf_map = block->functions_local_to_global();
+			auto local_nbf = bf_map.size();
+			int npoints = block->npoints();
+
+			auto veff_block = veff_->at(block->index());
+
+			int thread = 0;
+#ifdef _OPENMP
+			thread = omp_get_thread_num();
+#endif
+
+			pworkers[thread]->compute_functions(block);
+			auto phi_x = pworkers[thread]->basis_value("PHI_X");
+			auto phi_y = pworkers[thread]->basis_value("PHI_Y");
+			auto phi_z = pworkers[thread]->basis_value("PHI_Z");
+
+			// Preprocess kernel c²/(2c²-veff) * weight
+			double* w = block->w();
+			for (int p = 0; p < npoints; p++) {
+				kernel[p] = C2 /(2.*C2 - veff_block->get(p)) * w[p];
+			}
+
+			tmp->zero();
+
+			// Compute kinetic integral using kernel above
+			// T_SR --> non-relativistic T when veff --> 0.
+			// bf_map is needed because the basis funcions differ from that of a given block
+			for (int l_mu = 0; l_mu < local_nbf; l_mu++) {
+				int mu = bf_map[l_mu];
+				for (int l_nu = l_mu; l_nu < local_nbf; l_nu++) {
+					int nu = bf_map[l_nu];
+					for (int p = 0; p < npoints; p++) {
+						// ∇²φ(r)*kernel
+						tmp->add(mu,nu, kernel[p]*(
+							phi_x->get(p,l_mu)*phi_x->get(p,l_nu) +
+							phi_y->get(p,l_mu)*phi_y->get(p,l_nu) +
+							phi_z->get(p,l_mu)*phi_z->get(p,l_nu)
+						));
```

Comment:
```suggestion
						tmp_p[mu][nu] += (phi_x_p[p][l_mu] * phi_x_p[p][l_nu] + phi_y_p[p][l_mu] * phi_y_p[p][l_nu] + phi_z_p[p][l_mu] * phi_z_p[p][l_nu]) * kernel[p];
```

Normally I would not fuss about this but since this is in such a deep loop, using pointers rather than directly calling functions would make this operation a lot more efficient :)

Reply:
it is not surprising, since `add` is defined in the header of `matrix.h` as
```
    void add(int h, int m, int n, double val) {
#ifdef PSIDEBUG
        if (m > rowspi_[h] || n > colspi_[h ^ symmetry_]) {
            outfile->Printf("out of bounds: symmetry_ = %d, h = %d, m = %d, n = %d\n", symmetry_, h, m, n);

            throw PSIEXCEPTION("What are you doing, Rob?");
        }
#endif
        matrix_[h][m][n] += val;
    }
```
so it does get inlined.